platform/linux-generic/pktio/socket.c

```diff
@@ -855,6 +937,126 @@ static int sock_init_global(void)
 	return 0;
 }
 
+static int sock_recv_mq_tmo_select(pktio_entry_t * const *entry,
+				   const int index[],
+				   unsigned num_q, unsigned *from,
+				   odp_packet_t packets[], int num,
+				   uint64_t usecs, fd_set *readfds,
+				   int maxfd)
+{
+	struct timeval timeout;
+	unsigned i;
+	int ret;
+
+	timeout.tv_sec = usecs / 1000 / 1000;
+	timeout.tv_usec = usecs % (1000 * 1000);
+
+	for (i = 0; i < num_q; i++) {
+		ret = entry[i]->s.ops->recv(entry[i], index[i], packets, num);
+
+		if (ret > 0 && from)
+			*from = i;
+
+		if (ret != 0)
+			return ret;
+	}
+
+	if (select(maxfd + 1, readfds, NULL, NULL,
+		   usecs == ODP_PKTIN_WAIT ? NULL : &timeout) == 0)
+		return 0;
+
+	for (i = 0; i < num_q; i++) {
+		ret = entry[i]->s.ops->recv(entry[i], index[i], packets, num);
+
+		if (ret > 0 && from)
+			*from = i;
+
+		if (ret != 0)
+			return ret;
+	}
+
+	return 0;
+}
+
+int sock_recv_mq_tmo_try_int_driven(const struct odp_pktin_queue_t queues[],
+				    unsigned num_q, unsigned *from,
+				    odp_packet_t packets[], int num,
+				    uint64_t usecs, int *trial_successful)
+{
+	unsigned i;
+	pktio_entry_t *entry[num_q];
+	int index[num_q];
+	fd_set readfds;
+	int maxfd = -1;
+	int (*impl)(pktio_entry_t *entry[], int index[], int num_q,
+		    odp_packet_t packets[], int num, unsigned *from,
+		    uint64_t wait_usecs) = NULL;
+	int impl_set = 0;
+
+	/* First, we get pktio entries and queue indices. We then see if the
+	   implementation function pointers are the same. If they are the
+	   same, impl will be set to non-NULL; otherwise it will be NULL. */
+
+	for (i = 0; i < num_q; i++) {
+		entry[i] = get_pktio_entry(queues[i].pktio);
+		index[i] = queues[i].index;
+		if (entry[i] == NULL) {
+			ODP_DBG("pktio entry %d does not exist\n",
+				queues[i].pktio);
+			*trial_successful = 0;
+			return -1;
+		}
```

Comment:
By checking both ops here you can return right away and minimize overhead for pktios not supporting interrupt driven mode.
E.g.
```
		if (!entry[i]->s.ops->recv_mq_tmo && !entry[i]->s.ops->fd_set) {
			*trial_successful = 0;
			return -1;
		}
```

Reply:
I would prefer return 0; as the trial_successful already tells not to use this interrupt-driven way, but it probably doesn't matter much. I'll do this (with return 0).